common/archiver/s3store/util.go

```diff
@@ -113,9 +113,16 @@ func bucketExists(ctx context.Context, s3cli s3iface.S3API, URI archiver.URI) er
 }
 
 // Key construction
-func constructHistoryKey(path, domainID, workflowID, runID string, version int64) string {
+func constructHistoryKey(path, domainID, workflowID, runID string, version int64, batchIdx int) string {
+	prefix := constructHistoryKeyPrefixWithVersion(path, domainID, workflowID, runID, version)
+	// Note batchIdx has an upper limit of 200 and we rely on them being sorted alphabetically.
```

Comment:
May I ask what's the reason behind this upper limit?

Reply:
Sorry, 200 is way too low. I want to make sure the first version is returned first, so I would need to zero pad the number so 10 is sorted after 9. What would be a good upper bound for number of history batches? 9999?

Reply:
I understand you rely on the result being sorted when listing and you need a fixed length of the page number. But I guess my question is why do we need listing in the first place. We should be able to construct the entire key given the page number right? Then by keeping the page number in the next page token, we can also do pagination. Maybe I missed some context here?

Reply:
Yes, but then we would need to "test" if there is another page by calling exists on the next page item for each iteration of the loop. This would make it a non issue but less efficient

Reply:
I see. So for each page, we need to first call exist and then download? Is it possible to just call download and stop when encounter a not exist error?

Another way might be include some metadata information for each blob. For example, the upload dc, time, whether or not this is the last blob etc. This is what we do for our internal implementation.

The third option is just have a upper limit with 4 or 5 9s. But please document this limitation in the README.

I am ok with any approaches here, your choice.

Reply:
Went with using the metadata info that appears in HistoryBlobHeader, feels like the cleanest solution

Reply:
Awesome! Thanks for implementing this.